runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java

```diff
@@ -516,6 +519,25 @@ public DataflowPipelineJob run(Pipeline pipeline) {
 
     List<DataflowPackage> packages = options.getStager().stageFiles();
 
+    RunnerApi.Pipeline protoPipeline = PipelineTranslation.toProto(pipeline);
+    File serializedProtoPipeline;
+    try {
+      serializedProtoPipeline = File.createTempFile("pipeline", ".bin");
+      protoPipeline.writeDelimitedTo(new FileOutputStream(serializedProtoPipeline));
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    LOG.info("Staging pipeline description to {}", options.getStagingLocation());
+    DataflowPackage stagedPipeline;
+    try (PackageUtil packageUtil = PackageUtil.withDefaultThreadPool()) {
+      stagedPipeline =
+          packageUtil
+              .stageClasspathElements(
```

Comment:
How are we ensuring that we don't conflict the pipeline URL across jobs or other staged files? Do we do rely on a local temp file name and the MD5 suffix being added implicitly?